pkg/endpoint/policy.go

```diff
@@ -766,6 +611,15 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 
 	os.RemoveAll(backupDir)
 
+	// Update desired policy for endpoint because policy has now been realized
+	// in the datapath. PolicyMap state is not updated here, because that is
+	// performed in endpoint.syncPolicyMap().
+	if err == nil {
+		e.Mutex.Lock()
+		e.RealizedL4Policy = e.DesiredL4Policy
```

Comment:
Should this be done in the controller?

To me, `RealizedState` means that we have made a round-trip to the destination and received confirmation that the state is realized. Not that we've asynchronously pushed the request down.

Reply:
The problem is that the `L4Policy` object contains information that is realized (or not realized) in `regenerateBPF`, and the controller isn't aware of that at this time. Unless we want to put endpoint compilation / regeneration as well into a controller for an endpoint (which IMO would best be done in another patch), we should keep it here. Let me know if you want me to elaborate on my reasoning a bit more.

Reply:
Hmm. I see what you mean, but to be explicit let me attempt to explain my understanding and use that as a basis for discussion. Breaking down the `L4Policy`, it really represents three separate areas which can be realized in slightly differing manners:

* The L4 policy itself, at an L3/L4 layer. To implement this, we perform BPF map updates.
* The proxy redirect ports. We currently manage this through BPF code generation, though we intend to convert it into BPF map updates some time soon.
* Proxy updates via NPDS. 

So what you're saying is that `e.RealizedPolicy` represents that these updates have all been queued by the endpoint layer. The concern I have is that we lose introspection into what the latest known state of the lower layers is. From my reading of the current changes, I would only ever expect a difference between desired and realized state to exist for a very short period during regeneration.

What I'd expect instead is something like:
* Each area tracks its own policyRevision
* When the state is pushed successfully down, a callback updates the policyRevision for this area
* When the last of the three callbacks brings the three revisions in sync with each other, we update the realized state.

This way, when a successful push happens, we update the realized policy and it's clear what was implemented. When there's an error, the realizedpolicy would show you what is the most likely state that we're in, the last one to succeed (let's ignore the lack of #3897 for now). To know if the datapath is properly synced, you can compare the `endpoint get` output to see whether the desired and realized states are different..

Reply:
@joestringer I think that what you are saying is correct. However, I think that this is something that would best be done as a separate PR in the future. Right now, my approach is naive in that it waits until the datapath is fully realized, and only then updates the policy revision. I went with a more naive approach here because I want to approach this change incrementally instead of doing a huge refactor of everything all at once.

Reply:
I forgot about the fact that the thread performing this update will synchronously call the map sync function, so at least one attempt to update the BPF maps has been made when the policyRevision is updated. This is pretty much the same guarantee as today, so no regression. I agree that the above can be realized in a separate effort.

Reply:
@raybejjani can you elaborate?